### ISSUE TITLE
Generic self

### ISSUE THREAD

**Issue by Neppord** [vimeo/psalm#1282]
I would like to implement the functor pattern (oo style)

```
@template IN
@template OUT
@psalm-param callabel(IN):OUT $f
@psalm-return self<OUT>
```
Where IN sits on the interface and OUT on the map method.

I could hack this by adding a new template variable that points to the I inheriting class that then would be returned, like they do in java.

**Comment by muglug** [comment]
Reproduced with https://getpsalm.org/r/fcc3ad80cb, also includes generic `static`

**Comment by muglug** [comment]
This is a cool pattern! https://getpsalm.org/r/7fee89a6ca

**Comment by muglug** [comment]
`composer update vimeo/psalm` - I fixed and created a new release with the fix (3.0.14)

**Comment by muglug** [comment]
@Neppord run `composer update vimeo/psalm` - I fixed and created a new release with the fix (3.0.14)
